File: src/test/java/handler/FanduelHandlerTest.java
package handler;

import collect.dfs.Fanduel;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import util.AWSClient;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

class FanduelHandlerTest {

    @Mock
    Fanduel fanduel;

    @Mock
    AWSClient AWSClient;

    @InjectMocks
    FanduelHandler fanduelHandler;

    @BeforeEach
    void setUp() {
        MockitoAnnotations.initMocks(this);
        when(fanduel.getAllContestData("2019-12-16")).thenReturn(new ArrayList<>());
    }

    @Test
    void shouldHandleRequestForFanduelData() {
        Map<String, String> testMap = new HashMap<>();
        testMap.put("date", "2019-12-16");
        fanduelHandler.handleRequest(testMap);
        verify(fanduel).getAllContestData("2019-12-16");
        verify(AWSClient, never()).uploadToS3(anyString(), any());
    }

    @Test
    void shouldHandlePipelineRequestForFanduelData() {
        Map<String, String> testMap = new HashMap<>();
        testMap.put("invocationType", "pipeline");
        fanduelHandler.handleRequest(testMap);
        verify(AWSClient, times(1)).uploadToS3(anyString(), any());
    }
}

File: src/test/kotlin/handler/FanduelHandlerTest.kt
package handler

import io.mockk.*
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class FanduelHandlerTest {
    private val date = "a date"
    private val fanduelHandler = spyk<FanduelHandler>()
    private val fanduelResult: List<Map<String, Any>> = listOf()

    @BeforeEach
    fun setUp() {
        every { fanduelHandler.getFanduelContestData(date) } returns fanduelResult
        coJustRun { fanduelHandler.uploadToS3(any(), any()) }
    }

    @Test
    fun `should handle request with web invocation`() = runBlocking {
        val result = fanduelHandler.handleRequest(mapOf("date" to date))
        coVerify(exactly = 0) { fanduelHandler.uploadToS3(any(), any()) }
        assertEquals(fanduelResult, result)
    }

    @Test
    fun `should handle request with pipeline invocation`() = runBlocking {
        val result = fanduelHandler.handleRequest(mapOf("date" to date, "invocationType" to "pipeline"))
        coVerify { fanduelHandler.uploadToS3("fanduelData.json", fanduelResult) }
        assertEquals(listOf<Map<String, Any>>(), result)
    }
}